### PR TITLE
Don't try to clean up if child is killed gracefully

### PR DESCRIPTION
**Pull request by chrisbroome**
This is a potential fix for #5812 and has been tested on Windows 10. I'm mainly opening the PR to get feedback on the approach and start a bigger discussion around what the solution should be.

### CODE REVIEW

**Comment by packetzero** [comment]
It's nice to see some work on watcher kill.  It's far from graceful.  Looking at the source, I don't see any delay between killGracefully() and cleanup()/kill().  Sometimes on a busy system, with an intensive query, it can take a while to exit.  IMO, watcher should only be protecting against the runaway generate() function that never stops or eats up all resources, not the cases where usage spikes for a while.  It's not your responsibility to fix all these things here, I just wanted to vent since you are looking at it.

File: osquery/core/watcher.cpp [hunk 1]
```diff
@@ -345,21 +345,29 @@ bool WatcherRunner::watch(const PlatformProcess& child) const {
...
+    const auto message = std::string("Watcher cannot stop worker process (") + std::to_string(childPid) + ").";
+    LOG(INFO) << message;
+    return;
+  }
 
-  // Clean up the defunct (zombie) process.
-  if (!child.cleanup()) {
-    auto child_pid = child.pid();
+  const auto killedGracefully = child.killGracefully();
+  if (killedGracefully) {
```
Comment (uptycs-nishant) [inline osquery/core/watcher.cpp]: This fix is not correct for *nix systems.

bool PlatformProcess::killGracefully() const {
  if (!isValid()) {
    return false;
  }

  int status = ::kill(nativeHandle(), SIGINT);
  return (status == 0);
}

Sending a signal to a process on a *nix is async affair. So even if kill return successfully does not mean that process is terminated it only means that signal is delivered to that process so when the cpu scheduler picks up that process to run again it check if a signal pending for that process and based upon the signal disposition things happen( i.e. calling signal handler  or termination or ignoring signal)

File: osquery/core/watcher.cpp [hunk 2]
```diff
@@ -345,21 +345,29 @@ bool WatcherRunner::watch(const PlatformProcess& child) const {
...
+    LOG(INFO) << "Pid: " << childPid << " killed gracefully";
+    return;
+  }
 
-    LOG(WARNING) << "osqueryd worker (" << std::to_string(child_pid)
-                 << ") could not be stopped. Sending kill signal.";
+  LOG(WARNING) << "osqueryd worker (" << childPid << ") could not be stopped. Sending kill signal.";
```
Comment (directionless) [inline osquery/core/watcher.cpp]: You should add some kind of delay between the graceful kill, and the not graceful one.

File: osquery/core/watcher.cpp (hunk 1 above)
Comment (chrisbroome) [reply osquery/core/watcher.cpp]: Interesting. Thanks for the feedback! If there's so much difference between *nix systems and Windows does it make sense to use the same function to shutdown/cleanup? Maybe there should be 2 different implementations?

Comment (uptycs-nishant) [reply osquery/core/watcher.cpp]: Cleanup(child.cleanup()) is actually waiting(few seconds here in a loop) to confirm the termination of the child (worker) and reading its exit status  failure to do so will leave died worker process leave in zombie state unless its parent (watcher) dies as well.

**Comment by theopolis** [comment]
@chrisbroome, check out the changes merged from https://github.com/osquery/osquery/pull/6144. The shutdown logic changes may have improved this situation (just an assumption).

**Comment by chrisbroome** [comment]
@theopolis Thanks for taking a look at the shutdown logic in your PR. This PR was mainly just meant to spur a discussion and shed some light on the issue so that someone with way more knowledge of the shutdown process could actually look at it. Since that goal was achieved in #6144, I'll close this PR.

**Pull request closed without merging**